### PR TITLE
Error handling and `dry_run` to `docker/docker_nightly.py` and NVIDIA key sign bug fix

### LINKED ISSUE #1621: Improvements to `docker/build_image.sh`
`build_image.sh` has a few issues today

- [ ] The following command will actually build on CPU unless `-g` flag is also passed `./build_image.sh --cv=cu113`

- [x]  If `-g` flag is passed the below error is seen which is a known but yet unsolved issue https://github.com/NVIDIA/nvidia-docker/issues/1631

- [x] If this `docker/docker_nightly.py` fails it does so silently now, the Github action erroneously shows up as green even though if you go to dockerhub it's clear the builds aren't being published https://hub.docker.com/r/pytorch/torchserve-nightly/tags?page=1&name=gpu - something weird is going up because locally this script runs just fine

```
7 5.133 W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
#7 5.133 E: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease' is not signed.
```

- [ ] `kubernetes/kserve/build_image.sh` is not consistent with how it produces its build names and tags with the rest of our code

### PR DESCRIPTION
**Pull request by msaroufim**
Fixes #1621

### Added support for `dry_run`

We should do this for all our of scripts, it prints the commands that will be run without running them

```
(serve) ubuntu@ip-172-31-16-198:~/serve/docker$ python docker_nightly.py dry_run
Executing command: ./build_image.sh -bt dev -t pytorch/torchserve-nightly:cpu-2022.05.16
Executing command: ./build_image.sh -bt dev -g -cv cu102 -t pytorch/torchserve-nightly:gpu-2022.05.16
Executing command: docker push pytorch/torchserve-nightly:cpu-2022.05.16
Executing command: docker push pytorch/torchserve-nightly:gpu-2022.05.16
Executing command: docker tag pytorch/torchserve-nightly:cpu-2022.05.16 pytorch/torchserve-nightly:latest-cpu
Executing command: docker tag pytorch/torchserve-nightly:gpu-2022.05.16 pytorch/torchserve-nightly:latest-gpu
Executing command: docker push pytorch/torchserve-nightly:latest-cpu
Executing command: docker push pytorch/torchserve-nightly:latest-gpu
```

### Error handling code

Before this PR our Docker nightly build was succeeding for 2 weeks even though there were no docker images being published because of the #1624 

So instead we add some error handling code which will run the succesful commands without interruption but then keep track of all the failed commands and print them with an exception alongside a raise error so CI will correctly show up as red.

In the below example there were 2 failures

Docker permissions issue

```
The push refers to repository [docker.io/marksaroufim/torchserve-nightly]
5f70bf18a086: Preparing 
9612eab116d6: Preparing 
399399333b1b: Preparing 
7a83ed28d9b6: Preparing 
3e549931e024: Preparing 
denied: requested access to the resource is denied
```

And the NVIDIA key singing issue

```
#9 5.073 W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
#9 5.073 E: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease' is not signed.
------
executor failed running [/bin/sh -c apt-get update &&     DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y     fakeroot     ca-certificates     dpkg-dev     sudo     g++     git     python3.8     python3.8-dev     python3.8-venv     python3.8-distutils     build-essential     openjdk-11-jdk     curl     vim     && rm -rf /var/lib/apt/lists/*     && cd /tmp     && curl -O https://bootstrap.pypa.io/get-pip.py     && python3.8 get-pip.py]: exit code: 100
```

But now the error handling is correct

```
Traceback (most recent call last):
  File "docker_nightly.py", line 51, in <module>
    if failed_commands:
Exception: failed commands: ['./build_image.sh -bt dev -g -cv cu102 -t marksaroufim/torchserve-nightly:gpu-2022.05.16', 'docker push marksaroufim/torchserve-nightly:cpu-2022.05.16', 'docker push marksaroufim/torchserve-nightly:gpu-2022.05.16', 'docker tag marksaroufim/torchserve-nightly:gpu-2022.05.16 pytorch/torchserve-nightly:latest-gpu', 'docker push marksaroufim/torchserve-nightly:latest-cpu', 'docker push marksaroufim/torchserve-nightly:latest-gpu']
```

### NVIDIA key signing issue

Added to all our Dockerfiles which download something from NVIDIA explicitly

```
# Workaround from https://github.com/NVIDIA/nvidia-docker/issues/1632
RUN apt-key del 7fa2af80
RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/7fa2af80.pub
```

### CODE REVIEW

**Review by HamidShojanazeri — APPROVED** [review]

File: docker/Dockerfile.benchmark [hunk 1]
```diff
@@ -23,6 +23,11 @@ ARG LOCAL_BENCHMARK_SERVE_FOLDER
 
 ENV PYTHONUNBUFFERED TRUE
 
+# Workaround from https://github.com/NVIDIA/nvidia-docker/issues/1632
+RUN apt-key del 7fa2af80
+RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/3bf863cc.pub
+RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64/7fa2af80.pub
+
```
Comment (lxning) [inline docker/Dockerfile.benchmark:30]: [docker/Dockerfile.benchmark](https://github.com/pytorch/serve/pull/1627/files#diff-761111a87515028a7e5c334b926a2299484ef91217ae0a169ce5872937d72a8c) should be deprecated due benchmark automate deprecation.

Comment (msaroufim) [reply docker/Dockerfile.benchmark]: Done

File: docker/docker_nightly.py [hunk 2]
```diff
@@ -1,26 +1,52 @@
...
     today = date.today()
     return today.strftime("%Y.%m.%d")
 
 if __name__ == "__main__":
+
+    def try_and_handle(cmd, dry_run = False):
+        if dry_run:
+            print(f"Executing command: {cmd}")
+        else:
+            code = os.system(cmd)
```
Comment (lxning) [inline docker/docker_nightly.py:16]: will os.system return error if docker build fails? Should we use subprocess.run as https://github.com/pytorch/serve/pull/1623/files#diff-0377cd7efa39189d481ea81cd7c45057e164e01393c2316560bc6bf677b21517R17

Comment (msaroufim) [reply docker/docker_nightly.py:16]: `os.system(cmd)` will not return an error but it'll return a non zero error code if it fails and get added to a `failed_commands` list 

Then at the very end I raise an exception so the script will definitely fail and you can see this in the logs above in the description

```python
    if failed_commands:
        raise Exception(f"failed commands: {failed_commands}")
```

That said if you think `subprocess.run` is a better long term option I can use, `os.system()` is just easy

**Comment by msaroufim** [comment]
@lxning just changed to using `subprocess.run()` instead


### Dry run

```

(serve) ➜  docker git:(msaroufim-patch-13) ✗ python3 docker_nightly.py --organization marksaroufim --dry_run
Executing command: ./build_image.sh -bt dev -t marksaroufim/torchserve-nightly:cpu-2022.05.17
Executing command: ./build_image.sh -bt dev -g -cv cu101 -t marksaroufim/torchserve-nightly:gpu-2022.05.17
Executing command: docker push marksaroufim/torchserve-nightly:cpu-2022.05.17
Executing command: docker push marksaroufim/torchserve-nightly:gpu-2022.05.17
Executing command: docker tag marksaroufim/torchserve-nightly:cpu-2022.05.17 marksaroufim/torchserve-nightly:latest-cpu
Executing command: docker tag marksaroufim/torchserve-nightly:gpu-2022.05.17 marksaroufim/torchserve-nightly:latest-gpu
Executing command: docker push marksaroufim/torchserve-nightly:latest-cpu
Executing command: docker push marksaroufim/torchserve-nightly:latest-gpu
```


### Working example

```

(serve) ➜  docker git:(msaroufim-patch-13) ✗ python3 docker_nightly.py --organization marksaroufim
[+] Building 10.8s (9/13)                                                                                                                                                                          
 => [internal] load build definition from Dockerfile.dev                                                                                                                                      0.0s
 => => transferring dockerfile: 41B                                                                                                                                                           0.0s
 => [internal] load .dockerignore                                                                                                                                                             0.0s
 => => transferring context: 2B                                                                                                                                                               0.0s
 => resolve image config for docker.io/docker/dockerfile:experimental     
```

### Introduce a failure

`try_and_handle(f"./build_image.sh -bt dev -g -cv cu99 -t {organization}/{gpu_version}", dry_run)`

Then running

```
(serve) ➜  docker git:(msaroufim-patch-13) ✗ python3 docker_nightly.py --organization marksaroufim
CUDA version not supported
Traceback (most recent call last):
  File "docker_nightly.py", line 34, in <module>
    try_and_handle(f"./build_image.sh -bt dev -g -cv cu99 -t {organization}/{gpu_version}", dry_run)
  File "docker_nightly.py", line 18, in try_and_handle
    raise(e)
  File "docker_nightly.py", line 16, in try_and_handle
    subprocess.run([cmd], shell=True, check = True)
  File "/opt/homebrew/Cellar/python37/3.7.5_3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['./build_image.sh -bt dev -g -cv cu99 -t marksaroufim/torchserve-nightly:gpu-2022.05.17']' returned non-zero exit status 1.
```

File: docker/docker_nightly.py (hunk 2 above)
Comment (msaroufim) [reply docker/docker_nightly.py]: Updated to subprocess.run here https://github.com/pytorch/serve/pull/1627#issuecomment-1129467968

**Review by lxning — APPROVED** [review]

**Pull request merged**